spec/app_size_metrics_helper_spec.rb

```diff
@@ -0,0 +1,153 @@
+                                             'Group Metadata 2': 'Group Value 2'
+                                           })
+      metrics_helper.add_metric(name: 'Metric 1', value: 12_345, meta: { m1a: 'Metric 1 Metadata A' })
+      metrics_helper.add_metric(name: 'Metric 2', value: 67_890)
+      metrics_helper.add_metric(name: 'Metric 3', value: 13_579, meta: { m3a: 'Metric 3 Metadata A', m3b: 'Metric 3 Metadata B' })
+      metrics_helper
+    end
+    let(:expected_data) do
+      {
+        meta: [
+          { name: 'Group Metadata 1', value: 'Group Value 1' },
+          { name: 'Group Metadata 2', value: 'Group Value 2' },
+        ],
+        metrics: [
+          { name: 'Metric 1', value: 12_345, meta: [{ name: 'm1a', value: 'Metric 1 Metadata A' }] },
+          { name: 'Metric 2', value: 67_890 },
+          { name: 'Metric 3', value: 13_579, meta: [{ name: 'm3a', value: 'Metric 3 Metadata A' }, { name: 'm3b', value: 'Metric 3 Metadata B' }] },
+        ]
+      }.to_json
+    end
+
+    context 'when using file:// scheme for the URL' do
+      it 'writes the payload uncompressed to a file when disabling gzip' do
+        in_tmp_dir do |tmp_dir|
+          output_file = File.join(tmp_dir, 'payload.json')
+          file_url = File.join('file://localhost/', output_file)
+
+          code = metrics_helper.send_metrics(to: file_url, api_token: nil, use_gzip: false)
+
+          expect(code).to eq(201)
+          expect(File).to exist(output_file)
+          uncompressed_data = File.read(output_file)
+          expect(uncompressed_data).to eq(expected_data)
+        end
+      end
+
+      it 'writes the payload compressed to a file when enabling gzip' do
+        in_tmp_dir do |tmp_dir|
+          output_file = File.join(tmp_dir, 'payload.json.gz')
+          file_url = File.join('file://localhost/', output_file)
+
+          code = metrics_helper.send_metrics(to: file_url, api_token: nil, use_gzip: true)
+
+          expect(code).to eq(201)
+          expect(File).to exist(output_file)
+          uncompressed_data = Zlib::GzipReader.open(output_file, &:read)
+          expect(uncompressed_data).to eq(expected_data)
+        end
+      end
+    end
+
+    context 'when using non-file:// scheme for the URL' do
+      let(:api_url) { 'https://fake-metrics-server/api/grouped-metrics' }
+      let(:token) { 'fake#tokn' }
+
+      it 'sends the payload uncompressed to the server and with the right headers when disabling gzip' do
+        expected_headers = {
+          Authorization: "Bearer #{token}",
+          Accept: 'application/json',
+          'Content-Type': 'application/json'
+        }
+        last_received_body = nil
+        stub = stub_request(:post, api_url).with(headers: expected_headers) do |req|
+          last_received_body = req.body
+        end.to_return(status: 201)
+
+        code = metrics_helper.send_metrics(to: api_url, api_token: token, use_gzip: false)
+
+        expect(code).to eq(201)
```

Comment:
When I first read this, I questioned whether it was valuable to assert `code` is `201` when that's the value we set in `stub_request`.

But I came to the conclusion that it is indeed valuable: To ensure `send_metrics` returns the HTTP response status  code it received 👍

Reply:
Followup question after reading the production code this test verifies (I try to read the tests first and the production code after).

How do you feel about the fact that `201` is also the value returned by the file URI code path? Could it result in a false positive here? Or, given we rely on webmock which does a lot of extra checks for us, we can rest assured that if the file URI code path is hit instead of the API one we'll know about it? I'm leaning towards the latter 👍

Reply:
> So in fact that's fully intentional that this spec tests that the code returns 201, even for the cases when the file URI path is being tested.

Totally agree that it's appropriate to ensure `code` is `201` when _the file URI path_ is being tested 👍 

What I'm on the fence about is the `https://` behavior. This code is for the `https://` URI, though, and that's really what my question was about. 

The `https://` path [returns `response.code.to_i`](https://github.com/wordpress-mobile/release-toolkit/pull/364/files#diff-fabf5c60f39ba983a6eb9ba64e193689e04d4cab2694665dac19c8802f59b4afR91).

The tests stubs the response to return 201, the the same value that the `file://` code path would return. That's what made me wonder about the chance of a false positive, since, just by looking at `code`, I don't think we can which path the code took.

But "given we rely on webmock which does a lot of extra checks for us, we can rest assured that if the file URI code path is hit instead of the API one we'll know about it" because between webmock and the `expect(stub).to have_been_made.once` the test would fail. So, all in all, I think this is okay 👍 😄

Reply:
Ah I see.

But note that this `expect` is there not only to ensure that the status code was 201 (which we know because we stub it)… _but_ also (and mainly?) to check that the action [forwards that http status code (that the request / stub returned) as a return value of the action itself](https://github.com/wordpress-mobile/release-toolkit/pull/364/files#diff-fabf5c60f39ba983a6eb9ba64e193689e04d4cab2694665dac19c8802f59b4afR91) — as opposed to returning nothing, or the JSON payload, or whatnot as a return value.